tests/PySys/software-management-end-to-end/environment_sm_management.py

```diff
@@ -0,0 +1,217 @@
+
+import json
+import requests
+
+import pysys
+from pysys.basetest import BaseTest
+
+
+"""
+This environment provides an interface to software management features
+through the C8y REST API.
+With these we can emulate a user doing operations in the C8y UI.
+They are rather slow as they use the complete chain from end to end.
+
+WARNING: Handle with care!!!
+The C8YDEVICEID will handle on which this test will install and remove packages.
+"""
+
+
+def is_timezone_aware(stamp):
+    """determine if object is timezone aware or naive
+    See also: https://docs.python.org/3/library/datetime.html?highlight=tzinfo#determining-if-an-object-is-aware-or-naive
+    """
+    return stamp.tzinfo is not None and stamp.tzinfo.utcoffset(stamp) is not None
+
+
+class SmManagement(BaseTest):
+
+    def setup(self):
+        """Setup Environment"""
+
+        if self.myPlatform != "container":
+            self.skipTest("Testing the apt plugin is not supported on this platform")
+
+        tenant = self.project.tenant
+        user = self.project.username
+        password = self.project.c8ypass
+
+        # Place to save the id of the operation that we started.
+        # This is suitable for one operation and not for multiple ones running
+        # at the same time.
+        self.operation_id = None
+
+        auth = bytes(f"{tenant}/{user}:{password}", "utf-8")
+        self.header = {
+            b"Authorization": b"Basic " + base64.b64encode(auth),
+            b"content-type": b"application/json",
+            b"Accept": b"application/json",
+        }
+
+        # Make sure we have no last operations pending
+        self.wait_until_succcess()
+
+    def trigger_action(self, package_name, package_id, version, url, action):
+        """Trigger a installation or deinstallation of a package"""
+
+        self.trigger_action_json(
+            [
+                {
+                    "id": package_id,
+                    "name": package_name,
+                    "version": version,
+                    "url": url,
+                    "action": action,
+                }
+            ]
+        )
+
+    def trigger_action_json(self, json_content):
+        """Take an actions description that is then forwarded to c8y"""
+
+        url = "https://thin-edge-io.eu-latest.cumulocity.com/devicecontrol/operations"
+
+        payload = {
+            "deviceId": self.project.deviceid,
+            "description": "Apply software changes, triggered from PySys test",
+            "c8y_SoftwareUpdate": json_content,
+        }
+
+        req = requests.post(url, json=payload, headers=self.header)
+
+        jresponse = json.loads(req.text)
+
+        self.log.info(f"Response status: {req.status_code}")
+        self.log.info(f"Response to action: {json.dumps(jresponse, indent=4)}")
+        self.log.info(f"Started operation: {jresponse['id']}")
+
+        self.operation = jresponse
+        self.operation_id = jresponse["id"]
```

Comment:
This may raise an exxpecetion, either use try except block or `.get` and check for none.